### DISCUSSION TITLE
TextView & Highlighter performance discussion

### DISCUSSION THREAD

**Discussion by harshad1** [gsantner/markor#1749] [General] [answered]
- My markdown files are often many thousands of lines of various lists. 
- When highlighting is enabled, this effectively creates a ForegroundColorSpan for each line
    - I effectively have several thousand spans in my files
- EditText (SpannableStringBuilder) performance degrades pretty significantly with a large number of spans
    - This is an issue with the underlying android view
    - I have not found other text editor apps which don't suffer from this too

Note: The performance issue is not the time taken to highlight the file, but to _just type_ in a file with lots of highlighting

### I would like to brainstorm ideas for how to address it. So far here are the options I have come up with:

1. **Just disable highlighting** - This is the current approach. Performance on even very large files is sufficient with highlighting disabled (I don't see #1743 on my device)
2. **Enable selective highlighting** - Add options to disable list, bold, italics, strikeout highlighting etc etc. There is also some precedent for this  as we already have options for bigger headings and end of line highlighting.
3. **Find a way to improve performance of SpannableStringBuilder / EditText** - This will involve digging into the guts of edit text and I feel the risk of breaking things is too great
4. **See if there are performance improvements in AndroidX** - Another incentive to upgrade 😄

**Comment by gsantner** [comment]
What I see as only/best way to improve the performance is to only regex-scan and apply highlighting to screenarea (i.e. current visible lines ±3xscreenheight).

That being said, it is not straight forward - and at same time everything is still limited by TextViews general issues that appear also without any highlighting or something (really just plaintext). Loading 1-5 Megabytes of text is unfourtunatley not what textview is made for.

On uber view, realistically the only big performance gain would be to replace textview (&regex based) approach by something completley else. Which means basically throwing away everything.

**Comment by gsantner** [comment]
> See if there are performance improvements in AndroidX - Another incentive to upgrade smile

At least from TextView and EditText component side, I doubt it makes a relevant difference. At the end, both androidx EditText and TextView still inherit from the normal android.widget EditText & TextView components.
<https://androidx.tech/artifacts/appcompat/appcompat/1.4.2-source/androidx/appcompat/widget/AppCompatEditText.java.html>

**Comment by harshad1** [comment, marked as answer]
> apply highlighting to screenarea

I am going to attempt this. I think I have a way of doing it which is relatively straightforward.

**Reply by gsantner** [reply]
get current index, get line number of index, get index(line-100) + index(line+100), getText().subarea(-100, +100), regex.match()

something like that I guess?    
(Still that only can improve the performance at highlighting)

**Reply by harshad1** [reply]
Actually, since the immediate problem is typing editing being too slow when there are too many spans.

1. When the highlighter runs, save all spans to the HighlightingEditor. (an array of [span, startIndex, endIndex, type])
2. OnScroll find the indices currently on screen (+- 2 screens)
3. Go through list of spans and apply those which are needed
4. Remove those which are not needed

If this works and scrolling remains smooth, next step is to also compute the spans for the region in question.
